service/matching/version_sets.go

```diff
@@ -187,52 +150,86 @@ func hashBuildId(buildID string) string {
+		VersionSets: make([]*persistencespb.CompatibleVersionSet, len(data.GetVersionSets())),
 	}
 	copy(clone.VersionSets, data.GetVersionSets())
 	return &clone
 }
 
 func shallowCloneVersionSet(set *persistencespb.CompatibleVersionSet) *persistencespb.CompatibleVersionSet {
 	clone := &persistencespb.CompatibleVersionSet{
-		SetIds:                 set.SetIds,
-		BuildIds:               make([]*persistencespb.BuildId, len(set.BuildIds)),
-		DefaultUpdateTimestamp: set.DefaultUpdateTimestamp,
+		SetIds:                      set.SetIds,
+		BuildIds:                    make([]*persistencespb.BuildId, len(set.BuildIds)),
+		DefaultUpdateTimestamp:      set.DefaultUpdateTimestamp,
+		QueueDefaultUpdateTimestamp: set.QueueDefaultUpdateTimestamp,
 	}
 	copy(clone.BuildIds, set.BuildIds)
 	return clone
 }
 
+// UpdateVersionSets updates version sets given existing versioning data and an update request. The request is expected
+// to have already been validated.
+//
+// See the API docs for more detail. In short, the versioning data representation consists of a sequence of sequences of
+// compatible versions. Like so:
+//
+//	                     *
+//	┬─1.0───2.0─┬─3.0───4.0
+//	│           ├─3.1
+//	│           └─3.2
+//	├─1.1
+//	├─1.2
+//	└─1.3
+//
+// In the above example, 4.0 is the current default version and no other versions are compatible with it. The previous
+// compatible set is the 3.x set, with 3.2 being the current default for that set, and so on. The * represents the
+// current default set pointer, which can be shifted around by the user.
+//
+// A request may:
+//  1. Add a new version possibly as the new overall default version, creating a new set.
+//  2. Add a new version, compatible with some existing version, adding it to that existing set and making it the new
+//     default for that set.
+//  3. Target some existing version, marking it (and thus its set) as the default set.
+//
+// Deletions are performed by a background process which verifies build IDs are no longer in use and safe to delete (not yet implemented).
+//
+// Update may fail with FailedPrecondition if it would cause exceeding the supplied limits.
+func UpdateVersionSets(clock hlc.Clock, data *persistencespb.VersioningData, req *workflowservice.UpdateWorkerBuildIdCompatibilityRequest, maxSets, maxBuildIds int) (*persistencespb.VersioningData, error) {
+	data, err := updateImpl(clock, common.CloneProto(data), req)
+	if err != nil {
+		return nil, err
+	}
+	return data, checkLimits(data, maxSets, maxBuildIds)
+}
+
 //nolint:revive // cyclomatic complexity
-func updateImpl(timestamp hlc.Clock, existingData *persistencespb.VersioningData, req *workflowservice.UpdateWorkerBuildIdCompatibilityRequest) (*persistencespb.VersioningData, error) {
+func updateImpl(timestamp hlc.Clock, data *persistencespb.VersioningData, req *workflowservice.UpdateWorkerBuildIdCompatibilityRequest) (*persistencespb.VersioningData, error) {
```

Comment:
This should mention that it now mutates data, and possibly shouldn't return a pointer any more since that can imply it's creating a new thing

Reply:
It still needs to return a pointer because it might recreate the entire structure.

Reply:
Or data would have to be a double pointer.